File: y_minus_sim.c
#include <stdio.h>
#include <string.h>

// Minimal "Y-" Compiler Simulator
// This script simulates the logic of a 256-byte machine code compiler.

void emit(const char* name, const char* hex) {
    printf("%-15s : %s\n", name, hex);
}

void compile(const char* source) {
    printf("Compiling: %s\n", source);
    printf("-----------------------------------\n");

    const char* p = source;
    while (*p) {
        if (*p == ' ' || *p == '\n' || *p == '\t') { p++; continue; }

        // Variable assignment: x = 5;
        if (p[1] == '=') {
            char var = p[0];
            int offset = (var - 'a') * 4;
            p += 2; // skip 'x='

            if (p[0] >= '0' && p[0] <= '9') {
                // Constant assignment: x = 5;
                // MOV [RCX + offset], val -> C7 41 offset val
                char hex[32];
                sprintf(hex, "C7 41 %02X %02X 00 00 00", offset, p[0] - '0');
                emit("MOV_CONST", hex);
                p++;
            }
            if (*p == ';') p++;
        }
        
        // Return: r x;
        else if (*p == 'r') {
            p++; // skip 'r'
            while (*p == ' ') p++;
            char var = *p;
            int offset = (var - 'a') * 4;
            
            // MOV EAX, [RCX + offset] -> 8B 41 offset
            char hex[32];
            sprintf(hex, "8B 41 %02X", offset);
            emit("LOAD_EAX", hex);
            
            // RET -> C3
            emit("RET", "C3");
            p++;
            if (*p == ';') p++;
        }
        else {
            p++;
        }
    }
}

int main() {
    compile("a=5; r a;");
    return 0;
}

File: y_minus_sim.rs
fn main() {
    let source = "a=5; r a;";
    println!("Compiling: {}", source);
    println!("-----------------------------------");

    let mut chars = source.chars().peekable();
    while let Some(&c) = chars.peek() {
        if c.is_whitespace() {
            chars.next();
            continue;
        }

        if c.is_alphabetic() && chars.clone().nth(1) == Some('=') {
            let var = c;
            let offset = (var as u8 - b'a') * 4;
            chars.next(); // skip var
            chars.next(); // skip '='
            
            while let Some(&next) = chars.peek() {
                if next.is_whitespace() { chars.next(); } else { break; }
            }
            
            if let Some(&val_char) = chars.peek() {
                if val_char.is_digit(10) {
                    let val = val_char.to_digit(10).unwrap();
                    println!("MOV_CONST       : C7 41 {:02X} {:02X} 00 00 00", offset, val);
                    chars.next();
                }
            }
            if chars.peek() == Some(&';') { chars.next(); }
        } else if c == 'r' {
            chars.next(); // skip 'r'
            while let Some(&next) = chars.peek() {
                if next.is_whitespace() { chars.next(); } else { break; }
            }
            if let Some(&var) = chars.peek() {
                let offset = (var as u8 - b'a') * 4;
                println!("LOAD_EAX        : 8B 41 {:02X}", offset);
                println!("RET             : C3");
                chars.next();
            }
            if chars.peek() == Some(&';') { chars.next(); }
        } else {
            chars.next();
        }
    }
}
